ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableInsert.scala

```diff
@@ -160,13 +158,15 @@ private[events] trait EventsTableInsert { this: EventsTable =>
     * @throws RuntimeException If a value cannot be serialized into an array of bytes
     */
   @throws[RuntimeException]
-  def prepareBatchInsert(
+  def prepareBatchInsert[Nid <: NodeId](
```

Comment:
`NodeId` is a `final case class`, this is probably unnecessarily generic.

Reply:
Scala seems to think that `Nid` and `NodeId` are distinct, which is [exploitable in type-checking](https://typelevel.org/blog/2017/02/13/more-types-than-classes.html).

Reply:
To expand, there is no need for separate maps here at all, I would have rather mapped the party sets directly into the transaction structure. However, the transaction structure isn't sufficiently generic for that, so we have the nasty Perlish technique of "pass a map that associates the extra data in" instead.

Then I ask myself, "given that I don't want to hypergeneralize lf.Transaction, how can I make clear that these three arguments are related in a specific way, and they do not merely contain NodeIds, but NodeIds that are specifically related to each other?" And this is the most obvious tool available.

[I expand on a similar case for similarly final class String.](https://typelevel.org/blog/2017/09/05/three-types-of-strings.html)